### PR TITLE
Add user-level settings min_free_diskspace_bytes_to_throw_insert and min_free_diskspace_ratio_to_throw_insert

### LINKED ISSUE #65951: Add setting to stop INSERTs when disk space is close to 100%
**Use Case**

When ClickHouse fills the disk to 100%, it creates several negative side effects: inability to perform certain SELECT queries (which require temporary data), potential file corruption, inability to write logs, and other cascading issues that complicate recovery. It's preferable to stop writes slightly earlier to avoid these problems.

The same logic as `parts_to_throw_insert` can be applied here. Eventually, writes will fail as the disk becomes full; it's better if they fail slightly earlier to prevent cascading problems.

**Describe the solution you'd like**

Introduce user-level settings: `min_free_diskspace_bytes_to_throw_insert ` (default 10GB) and `min_free_diskspace_ratio_to_throw_insert ` (default 0.02, i.e., 2%). When these settings are non-zero, during an INSERT operation, ClickHouse will ensure that the storage policy of the target table has at least `min_free_diskspace_bytes_to_throw_insert` bytes or `min_free_diskspace_ratio_to_throw_insert` free space available. If not, it will throw an exception.

**Describe alternatives you've considered**

Relying on user discipline, monitoring, and alerting has proven to be ineffective.

**Additional context**

Throttling / delaying / slowing down the inserts in that conditions is more controversial, so I don't think it's really needed.

May be we can also make the logic a bit more complex - by checking if the default drive have that amount of free disks space AND there is at least one disk in storage policy of the table with that amount of the free disk space.

### PR DESCRIPTION
**Pull request by marco-vb**
Resolves #65951 

Added two new user-level settings:
- `min_free_disk_bytes_to_throw_insert`
- `min_free_disk_ratio_to_throw_insert`

These prevent inserts from occurring if disk is close to full, but do not affect operations that need to write temporary data to disk. See #65951 discussion to better understand why this is useful.

### Changelog category (leave one):
- Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Added user-level settings `min_free_disk_bytes_to_throw_insert` and `min_free_disk_ratio_to_throw_insert` to prevent insertions on disks that are almost full.

### Documentation entry for user-facing changes

- [x] Documentation is written (mandatory for new features)

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

#### CI Settings (Only check the boxes if you know what you are doing):
- [ ]  Allow: All Required Checks
- [ ]  Allow: Stateless tests
- [ ]  Allow: Stateful tests
- [ ]  Allow: Integration Tests
- [ ]  Allow: Performance tests
- [ ]  Allow: All Builds
- [ ]  Allow: batch 1, 2 for multi-batch jobs
- [ ]  Allow: batch 3, 4, 5, 6 for multi-batch jobs
---
- [ ]  Exclude: Style check
- [ ]  Exclude: Fast test
- [ ]  Exclude: All with ASAN
- [ ]  Exclude: All with TSAN, MSAN, UBSAN, Coverage
- [ ]  Exclude: All with aarch64, release, debug
---
- [ ]  Run only fuzzers related jobs (libFuzzer fuzzers, AST fuzzers, etc.)
- [ ]  Exclude: AST fuzzers
---
- [ ]  Do not test
- [ ]  Woolen Wolfdog
- [ ]  Upload binaries for special builds
- [ ]  Disable merge-commit
- [ ]  Disable CI cache

### CODE REVIEW

**Comment by marco-vb** [comment]
hey @jsc0218 thank you for adding the test label, this pr is now ready for review if you want to take a look!

**Comment by jsc0218** [comment]
@marco-vb Everything looks good. Just waiting for the private CI to pass. I would handle that part if there is some error.

File: src/Core/Settings.h [hunk 1]
```diff
@@ -343,6 +343,8 @@ class IColumn;
     M(Int64, distributed_ddl_task_timeout, 180, "Timeout for DDL query responses from all hosts in cluster. If a ddl request has not been performed on all hosts, a response will contain a timeout error and a request will be executed in an async mode. Negative value means infinite. Zero means async mode.", 0) \
     M(Milliseconds, stream_flush_interval_ms, 7500, "Timeout for flushing data from streaming storages.", 0) \
     M(Milliseconds, stream_poll_timeout_ms, 500, "Timeout for polling data from/to streaming storages.", 0) \
+    M(UInt64, min_free_disk_bytes_to_throw_insert, 0, "Minimum free disk space bytes to throw an insert.", 0) \
```
Comment (Algunenano) [inline src/Core/Settings.h:346]: This is not doing anything at all.

**Review by Algunenano — COMMENTED** [review]
The settings are declared both as common settings and mergetree settings but the common/context settings are never used.

This does not cover https://github.com/ClickHouse/ClickHouse/issues/65951 at all. Let's revert and reintroduce with proper tests

**Comment by marco-vb** [comment]
hey @Algunenano thanks for the feedback!

>The settings are declared both as common settings and mergetree settings but the common/context settings are never used.
You are absolutely correct, I eventually noticed this but forgot to remove it from Core settings.

>This does not cover #65951 at all.
What do you mean by this? Aside the unneeded settings in Core, which part of the issue was left uncovered? I am very new to this project and this may not be a good implementation but from manual and integration testing it seemed to me that it behaved as expected.

**Comment by Algunenano** [comment]
> What do you mean by this? Aside the unneeded settings in Core, which part of the issue was left uncovered?

The problem is that the settings were not unneeded in core. Without them being implemented properly you can't change the setting in user profiles, sessions or in standalone queries; you can only set them in the mergetree table settings (which makes them not easy to change).

**Comment by marco-vb** [comment]
Ok will take a look into that. Also, you say this is not properly tested -- what additional behavior would you like to see tested?

**Comment by jsc0218** [comment]
My mistake overlooked the change in the core setting. Indeed, it would be more reasonable in the core setting.

**Comment by marco-vb** [comment]
Thank you both for feedback @jsc0218 @Algunenano 
Submitted [improved PR](https://github.com/ClickHouse/ClickHouse/pull/69755) which hopefully addresses everything requested!

**Pull request merged**